crates/red_knot_python_semantic/resources/mdtest/terminal_statements.md

```diff
@@ -0,0 +1,638 @@
+the `raise` are not visible in any `else` clause, but are visible after the containing `try`
+statement, since `try` does not introduce a new scope.
+
+TODO: We are not currently implementing the "jump" behavior correctly for `raise` statements. The
+false positives in this section are because of that, and not our terminal statement support.
+
+```py
+def raise_in_then_branch(cond: bool):
+    x = "before"
+    try:
+        if cond:
+            x = "raise"
+            reveal_type(x)  # revealed: Literal["raise"]
+            raise ValueError
+        else:
+            x = "else"
+            reveal_type(x)  # revealed: Literal["else"]
+        reveal_type(x)  # revealed: Literal["else"]
+    except ValueError:
+        # TODO: Literal["raise"]
+        reveal_type(x)  # revealed: Literal["before", "raise", "else"]
+    except:
+        # TODO: Literal["raise"] or Never
+        reveal_type(x)  # revealed: Literal["before", "raise", "else"]
+    else:
+        reveal_type(x)  # revealed: Literal["else"]
+    finally:
+        # This includes "before" because we assume that an exception might have occurred before the
+        # `if` statement.
+        reveal_type(x)  # revealed: Literal["before", "raise", "else"]
+    reveal_type(x)  # revealed: Literal["before", "raise", "else"]
+
+def raise_in_else_branch(cond: bool):
+    x = "before"
+    try:
+        if cond:
+            x = "else"
+            reveal_type(x)  # revealed: Literal["else"]
+        else:
+            x = "raise"
+            reveal_type(x)  # revealed: Literal["raise"]
+            raise ValueError
+        reveal_type(x)  # revealed: Literal["else"]
+    except ValueError:
+        # TODO: Literal["raise"]
+        reveal_type(x)  # revealed: Literal["before", "else", "raise"]
+    except:
+        # TODO: Literal["raise"] or Never
+        reveal_type(x)  # revealed: Literal["before", "else", "raise"]
+    else:
+        reveal_type(x)  # revealed: Literal["else"]
+    finally:
+        # This includes "before" because we assume that an exception might have occurred before the
+        # `if` statement.
+        reveal_type(x)  # revealed: Literal["before", "else", "raise"]
+    reveal_type(x)  # revealed: Literal["before", "else", "raise"]
+
+def raise_in_both_branches(cond: bool):
+    x = "before"
+    try:
+        if cond:
+            x = "raise1"
+            reveal_type(x)  # revealed: Literal["raise1"]
+            raise ValueError
+        else:
+            x = "raise2"
+            reveal_type(x)  # revealed: Literal["raise2"]
+            raise ValueError
+    except ValueError:
+        # TODO: Literal["raise1", "raise2"]
+        reveal_type(x)  # revealed: Literal["before", "raise1", "raise2"]
+    except:
+        # TODO: Literal["raise1", "raise2"] or Never
+        reveal_type(x)  # revealed: Literal["before", "raise1", "raise2"]
+    else:
+        # This is unreachable
+        pass
```

Comment:
I realize we are intentionally not asserting about types _within_ unreachable code, but can we still verify that we consider this code unreachable by adding an assignment to `x` here and verifying it is not a possible value for `x` in the end?

Reply:
Done